Fix device discovery through iSCSI for SoftLayer

bgandon opened the pull request:
iSCSI device discovery could not be called a second time after disk partioning,
since code was refactored for IaaS-native disk resize.
Here we modify slightly the algorithm so that a never-mounted but
already-partitioned disk doesn't trigger any unnecessary iSCSI restart.

Fixes #252.

While working on this fix, I've reviewed other strategy classes implementing the `DevicePathResolver` interface. Conclusion is that only the iSCSI strategy is dependant on whether the disk device is partitioned or not, which leads to different results when called before or after partitioning.

rkoster commented:
@donacarr Could you verify this change indeed fixes #252?

friegger commented:
Looks good to me, let's wait for feedback from @donacarr.

@bgandon Out of curiosity, is iSCSI only used on SoftLayer?

donacarr commented:
The fix is partially working... I put comment here #252

donacarr commented:
We need still the fix.
Thx

bgandon commented:
Hello @donacarr, as discussed last month I've implemented a small cache in order to fix the issue pointed in your https://github.com/cloudfoundry/bosh-agent/issues/252#issuecomment-977712657. I've also added more unit tests in order to cover the issue that had been identified.

Could you please give it a shot and provide feedback? Thank you!

rkoster commented:
@donacarr were you able to test the changes made by Benjamin?

donacarr commented:
I plan to test it in the next two weeks

bgandon commented:
@donacarr, any new on testing the new updated code?

bgandon commented:
@donacarr, any new on testing the new updated code?

donacarr commented:
Hi @bgandon , sorry for late answer ... I was sick..
Btw my collegue tried your updated code but he got a failure:

```
Failures:

  1) with release and stemcell and subsequent deployments with persistent disk size changing should migrate disk contents
     Failure/Error: expect(persistent_disk('colocated', 0, deployment: deployment)).to_not eq(@size)
       
       expected: value != "20509264"
            got: "20509264"
       
       (compared using ==)
     # ./spec/system/with_release_stemcell_spec.rb:103:in `block (3 levels) in <top (required)>'
```

Unfortunately he did not collect new logs, so if you need it then the next time we build new stemcell, we will collect it...

bgandon commented:
Hello @donacarr,
With the issue you've encountered, it seems that growing the disk size using the classical disk migration process has failed in the Bosh Acceptance Tests.
It's not easy to relate that with the fix in this PR for resolving iSCSI device paths.
Yes, we would need the Agent logs in order to investigate the problem that prevented the disk from being properly resized.
Could you please provide these? Thank you.

rkoster commented:
@bgandon could you rebase these changes against master?

rkoster commented:
ping @bgandon

bgandon commented:
Sorry for late reply, will rebase soon!

bgandon commented:
All right, it's rebased!

jpalermo commented:
Has this been validated to fix the problem? @donacarr created the original issue, but I haven't seen confirmation that these changes fix things.

donacarr commented:
@jpalermo yes the issue was fixed by the new code.
thank you

jpalermo reviewed (approved):
Seems good to me

ramonskie reviewed (approved):
LGTM
